### ISSUE TITLE
Improve engine garbage collection

### ISSUE THREAD

**Issue by sipsma** [dagger/dagger#3849]
The current approach for ensuring we don't leave a bunch of engine images+binaries around running forever is very primitive and doesn't allow multiple versions of an engine to coexist. More background in this comment and the one preceding it: https://github.com/dagger/dagger/pull/3647#issuecomment-1302721770

This means that we must always update engine dependencies in SDKs very close to each other and that users must update those SDKs in lockstep too if they use multiples from the same machine.

In the long term, we can improve backwards compatibility so that SDKs don't need to be so tied to a single engine version. Additionally, other provisioning targets besides docker may offer different ways of solving this problem.

However, in the short-to-medium term we can improve this situation with a moderate amount of effort on something we likely want in general anyways:
1. We can add support for shutting down the runner if no active connections have been received for a given amount of time. This is something @shykes has mentioned previously and has actually been discussed recently in a buildkitd issue: https://github.com/moby/buildkit/issues/3252#issuecomment-1300987965
2. This approach will result in engine containers stopping after they haven't been used, which makes them available for cleanup by the user via `docker system prune`. I personally would be okay with that since running a single prune command periodically (something many-to-most docker users do anyways) is not much to ask in exchange for having the ability to use SDKs with different engine versions.

**Comment by jedevc** [comment]
Agreed with point 1, whatever we do ends up needing some mechanism for that.

Point 2 though, does introduce a new point of friction for users - I'm a bit worried that this would be an annoying breaking change for most users.

*Ideally*, we'd have a way for engines to tear themselves down after not having been used for a period of time. But I can't see a way to convince the daemon to do this - maybe if we had a `--restart=always` or similar, then the engine could wipe it's own volume (even though it can't actually be deleted without propagation of the docker socket, which is annoying).

Not really sure, but surely anything has to be better than the current setup 😢
